core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -0,0 +1,263 @@
+  private final Object lock = new Object();
+
+  private Listener listener;
+  /** 'lock' must be held when assigning to delegate. */
+  private volatile ClientTransport delegate;
+
+  @GuardedBy("lock")
+  private Collection<PendingStream> pendingStreams = new HashSet<PendingStream>();
+  @GuardedBy("lock")
+  private Collection<PendingPing> pendingPings = new ArrayList<PendingPing>();
+  /**
+   * When shutdown == true and pendingStreams == null, then the transport is considered terminated.
+   */
+  @GuardedBy("lock")
+  private boolean shutdown;
+
+  @Override
+  public void start(Listener listener) {
+    this.listener = Preconditions.checkNotNull(listener, "listener");
+  }
+
+  @Override
+  public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata headers) {
+    ClientTransport transport = delegate;
+    if (transport != null) {
+      return transport.newStream(method, headers);
+    }
+    synchronized (lock) {
+      // Check again, since it may have changed while waiting for lock
+      transport = delegate;
+      if (transport != null) {
+        return transport.newStream(method, headers);
+      }
+      if (!shutdown) {
+        PendingStream pendingStream = new PendingStream(method, headers);
+        pendingStreams.add(pendingStream);
+        return pendingStream;
+      }
+    }
+    DelayedStream stream = new DelayedStream();
+    stream.setError(Status.UNAVAILABLE.withDescription("transport shutdown"));
+    return stream;
+  }
+
+  public void ping(final PingCallback callback, Executor executor) {
+    ClientTransport transport = delegate;
+    if (transport != null) {
+      transport.ping(callback, executor);
+      return;
+    }
+    synchronized (lock) {
+      // Check again, since it may have changed while waiting for lock
+      transport = delegate;
+      if (transport != null) {
+        transport.ping(callback, executor);
+        return;
+      }
+      if (!shutdown) {
+        PendingPing pendingPing = new PendingPing(callback, executor);
+        pendingPings.add(pendingPing);
+        return;
+      }
+    }
+    executor.execute(new Runnable() {
+        @Override public void run() {
+          callback.onFailure(
+              Status.UNAVAILABLE.withDescription("transport shutdown").asException());
+        }
+      });
+  }
+
+  /**
+   * Prevents creating any new streams until {@link #setTransport} is called. Buffered streams are
+   * not failed, so if {@link #shutdown} is called when {@link #setTransport} has not been called,
+   * you still need to call {@link setTransport} to make this transport terminated.
+   */
+  @Override
+  public void shutdown() {
+    boolean shouldCallTerminated = false;
+    synchronized (lock) {
+      if (shutdown) {
+        return;
+      }
+      shutdown = true;
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        pendingStreams = null;
+        shouldCallTerminated = true;
+      }
+    }
+    listener.transportShutdown(
```

Comment:
The listeners really will need to be called while the lock is held. Otherwise you can have a `transportTerminated()` be called simultaneously or even before `transportShutdown`, due to something like `PendingStream.setError()`.

Holding `lock` a little longer for the callbacks is no big deal. Normally the lock would be held while doing HTTP/2 processing. Since this "transport" only holds the lock for stream creation and deletion, the lock should be relatively uncontested compared to a "real" transport.